File: backend/schemas/tasks.py
from pydantic import BaseModel, Field, validator
from typing import List, Optional, Dict
from datetime import datetime
from models.models import Platform

class SearchParams(BaseModel):
    """搜索参数模型"""
    keywords: List[str]
    location: Optional[List[str]] = None
    min_followers: Optional[int] = None
    max_followers: Optional[int] = None
    min_following: Optional[int] = None
    max_following: Optional[int] = None
    min_posts: Optional[int] = None
    max_posts: Optional[int] = None
    is_verified: Optional[bool] = None
    is_private: Optional[bool] = None
    has_website: Optional[bool] = None
    category: Optional[str] = None

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "example": {
                "keywords": ["fashion", "style"],
                "location": ["New York"],
                "min_followers": 1000,
                "max_followers": 100000,
                "is_verified": True
            }
        }
    }

class SearchTaskCreate(BaseModel):
    """创建搜索任务请求模型"""
    name: str = Field(..., min_length=2, max_length=50)
    platform: Platform
    search_params: SearchParams
    results_limit: Optional[int] = Field(default=1000, ge=20, le=10000)

    model_config = {
        "from_attributes": True
    }

    @validator('platform')
    def validate_platform(cls, v):
        if isinstance(v, str):
            platform = Platform._missing_(v)
            if platform is None:
                raise ValueError('无效的平台类型')
            return platform
        return v

class SearchTaskUpdate(BaseModel):
    status: Optional[str] = None
    result_count: Optional[int] = None
    is_completed: Optional[bool] = None
    error_message: Optional[str] = None

    model_config = {
        "from_attributes": True
    }

class SearchTaskResponse(BaseModel):
    id: int
    name: str
    platform: Platform
    search_params: Dict
    status: str
    result_count: int
    results_limit: int
    is_completed: bool
    error_message: Optional[str]
    created_at: datetime
    completed_at: Optional[datetime]
    type: str = 'search'

    model_config = {
        "from_attributes": True,
        "json_encoders": {
            datetime: lambda v: v.isoformat()
        }
    } 

File: frontend/src/services/tasks.ts
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL;

export interface SearchTaskFilters {
    minFollowers?: number;
    maxFollowers?: number;
    isVerified?: boolean;
    isPrivate?: boolean;
}

export interface SearchTaskParams {
    platform: string;
    keywords: string;
    filters: SearchTaskFilters;
}

export interface SearchTask {
    id: number;
    name: string;
    platform: string;
    search_params: {
        keywords?: string[];
        location?: string[];
        min_followers?: number;
        max_followers?: number;
        min_following?: number;
        max_following?: number;
        min_posts?: number;
        max_posts?: number;
        is_verified?: boolean;
        is_private?: boolean;
        has_website?: boolean;
        category?: string;
    };
    status: 'pending' | 'processing' | 'completed' | 'failed';
    result_count: number;
    results_limit: number;
    is_completed: boolean;
    error_message?: string;
    created_at: string;
    completed_at?: string;
    type: string;
}

export interface SearchTaskListResponse {
    data: SearchTask[];
    total: number;
    page: number;
    page_size: number;
}

export const searchTaskService = {
    // 创建搜索任务
    async createTask(params: SearchTaskParams): Promise<SearchTask> {
        const response = await axios.post(`${API_URL}/search-tasks`, params);
        return response.data;
    },

    // 获取任务列表
    async getTasks(params?: { platform?: string; page?: number; pageSize?: number }): Promise<SearchTaskListResponse> {
        const { platform, page = 1, pageSize = 10 } = params || {};
        
        const response = await axios.get(`${API_URL}/search-tasks`, {
            params: {
                platform,
                page,
                pageSize
            }
        });
        return response.data;
    },

    // 获取任务详情
    async getTask(taskId: number): Promise<SearchTask> {
        const response = await axios.get(`${API_URL}/search-tasks/${taskId}`);
        return response.data;
    },

    // 删除任务
    async deleteTask(taskId: number): Promise<void> {
        await axios.delete(`${API_URL}/search-tasks/${taskId}`);
    },

    // 轮询任务状态
    async pollTaskStatus(taskId: number, callback: (task: SearchTask) => void, interval = 3000): Promise<void> {
        const checkStatus = async () => {
            const task = await this.getTask(taskId);
            callback(task);
            
            if (!task.is_completed && task.status !== 'failed') {
                setTimeout(checkStatus, interval);
            }
        };
        
        await checkStatus();
    }
}; 